### ISSUE TITLE
Light/dark theme disconnected from light/dark scheme

### ISSUE THREAD

**Issue by posita** [squidfunk/mkdocs-material#2727]
## Expected behavior

After reading [this](https://github.com/squidfunk/mkdocs-material/blob/4247e3b440644cfcf7219f1b32e147f13a490f13/docs/setup/changing-the-colors.md#system-preference), as well as squidfunk/mkdocs-material#1266, my understanding is:

1. Toggling the theme using the toggle switch will switch the preferred color scheme; and
1. Switching the preferred color scheme will toggle the theme.

However, they are disconnected as far as I can tell.

## Observed behavior

1. Toggling the theme using the switch in the header of this page will update the theme, but not the scheme.
1. Toggling the preferred color scheme (e.g., on MacOS, switching from `Light` to `Dark` in `System Preferences` » `General` » `Appearance`) will update the scheme, but not the theme.

## Detail

I have created [an example](https://posita.github.io/mkdocs-dark-bug/) that illustrates this. (Source can be found at [**posita/mkdocs-dark-bug**](https://github.com/posita/mkdocs-dark-bug).)

Test matrix:

| Theme&nbsp;👉<br />Scheme&nbsp;👇 | Light | Dark  |
| --------: | :---: | :---: |
| **Light** | ![Light/Light](https://posita.github.io/mkdocs-dark-bug/latest/light-theme-light-scheme.png) | ![Dark/Light](https://posita.github.io/mkdocs-dark-bug/latest/dark-theme-light-scheme.png) |
| **Dark**  |  ![Light/Dark](https://posita.github.io/mkdocs-dark-bug/latest/light-theme-dark-scheme.png)  |  ![Dark/Dark](https://posita.github.io/mkdocs-dark-bug/latest/dark-theme-dark-scheme.png)  |

Tested with:

| package                    | version |
| :------------------------- | ------: |
| mkdocs                     | 1.1.2   |
| mkdocs-material            | 7.1.6   |
| mkdocs-material-extensions | 1.0.1   |

**Comment by squidfunk** [comment]
Thanks for reporting. `prefers-color-scheme` uses the system preference, correct. Material for MkDocs can only react to system preference, but not change it. That’s just not possible. Thus, everything is working as expected.

**Comment by posita** [comment]
But that's partially the point: MkDocs is _**not**_ reacting to the system change. Only my image is. Consider the first column in my test matrix above. I am switching the system them. _**Only**_ the image changes. The MkDocs theme does not. The MkDocs them only changes when I explicitly toggle it.

**Comment by squidfunk** [comment]
If we would tie color schemes to system preferences to achieve what you’re asking for, it would not be possible to toggle colors. Only the system preference would be allowed. It’s not possible to achieve flexible colors schemes.

**Comment by squidfunk** [comment]
Let me rephrase: I don’t know how the functionality you’re asking for should work. Maybe I’m just not seeing it. If you can craft a PR that keeps all existing functionality, and additionally achieves what you’re aiming for, I’m more than happy to merge it.

**Comment by posita** [comment]
All right, maybe this is a documentation bug then? What does [this section](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/#system-preference) actually mean? This (**_emphasis_** mine), in particular, is quite confusing (at least to me):

> **_In order to automatically set the color palette to the user's system preference_**, a media query can be set as part of the `media` field next to the toggle definition in `mkdocs.yml` … .

That seems to strongly suggest the two are related in some way. I could see it meaning one of two things:

1. MkDocs-material reacts to the system preferences and then toggles the theme accordingly (the most reasonable interpretation); or
1. MkDocs-material overrides the scheme for its own pages when the theme is manually toggled.

What you're saying here is that they're not related at all, and that's by design? As currently written, it's unclear based on its text what the point of that section/configuration is if it doesn't mean one of those two things. What am I missing? What does “… automatically set the color palette to the user's system preference …” refer to?

**Comment by squidfunk** [comment]
1. If the user didn’t set the scheme explicitly, it is sourced from the system preference (or whatever you use as media queries)
2. When the user toggled the scheme, this setting is used

Thus, both are correct.

**Comment by posita** [comment]
Ah. Okay. I think understand now. I think there are two issues here (perhaps one documentation bug and one feature request):

1. Where is the behavior you outlined in #issuecomment-851706120 documented? If it's not, I'd suggest documenting it. If (or once) it is documented, I would suggest linking to that from [that section](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/#system-preference) to provide the requisite context.
1. The behavior you describe (of detecting once, and then persisting indefinitely) means that the site will not react to subsequent changes to the system preferences. I would suggest either ignoring them or reacting to them, but not once one way and then the other. This can lead to unintuitive behaviors. For example if the user's system preferences are set to change scheme automatically (e.g., "light" during daylight hours, "dark" otherwise), it means the MkDocs-material theme will "stick" to something depending on the time of day the user first came to it, which seems arbitrary.

Also, this came about by me trying to match dark/light images to dark/light themes. Irrespective of system preferences, how does one do that? It's not clear from the [Changing the Colors](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/) section. (Unless I just completely missed it?)

**Comment by posita** [comment]
A couple more questions:

1. Can you point me to where/how the theme information get persisted? I'll consider crafting a PR, but a pointer to that mechanism would be helpful.
1. This came about by me trying to match dark/light images to dark/light themes. Irrespective of system preferences, how does one do that? It's not clear from the [Changing the Colors](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/) section. (Unless I just completely missed it?)

**Comment by squidfunk** [comment]
Sure, here are the relevant parts:

https://github.com/squidfunk/mkdocs-material/blob/86f49eb21c6e7343875ba0d535b8fb38711b73a1/src/partials/javascripts/palette.html#L23-L29

https://github.com/squidfunk/mkdocs-material/blob/86f49eb21c6e7343875ba0d535b8fb38711b73a1/src/assets/javascripts/components/palette/index.ts#L78-L108

I suggest you create a discussion to discuss 2. (or ask on StackOverflow), as this is beyond the scope of this project.

**Comment by squidfunk** [comment]
Haven't seen your previous comment. 

> Where is the behavior you outlined in the previous comment documented? If it's not, I'd suggest documenting it. If (or once) it is documented, I would suggest linking to that from that section to provide the requisite context.

[From the docs](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/#system-preference):

> When the user __first visits your site__, the media queries are evaluated in the order of their definition. The first media query that matches selects the default color palette.

I really don't want to make it any more complicated than that. We've had other requests for clarification of the documentation in the past, but normally it got a lot more complicated leading to more confusion among users. If you feel that changing 2-3 words will yield better results, I invite you to go ahead and create a PR.

> The behavior you describe (of detecting once, and then persisting indefinitely) means that the site will not react to subsequent changes to the system preferences. 

Correct. This is currently unsupported.

**Comment by posita** [comment]
> I really don't want to make it any more complicated than that. We've had other requests for clarification of the documentation in the past, but normally it got a lot more complicated leading to more confusion among users. If you feel that changing 2-3 words will yield better results, I invite you to go ahead and create a PR.

Fair, but this issue is signal that at least some confusion already exists. I have no doubts that the current wording is an accessible reminder to someone who was already familiar with the subtleties, but that person is a distinctly different customer than one leaning heavily on the documentation to achieve that familiarity. In fairness, I accept that I may be the sole person to encounter this confusion, but if anyone else voices similar confusion, consider revisiting.

In the mean time, I will consider how to address my concerns in prose and submit a PR if I settle on something worthwhile.

As an aside, my hunch is that `toggle` and `media` should probably be treated as mutually exclusive (at least in practice). This conclusion stems from my prior observation that:

> _[The [documented configuration for system preferences](https://squidfunk.github.io/mkdocs-material/setup/changing-the-colors/#system-preference)]_ can lead to unintuitive behaviors. For example if the user's system preferences are set to change scheme automatically (e.g., "light" during daylight hours, "dark" otherwise), it means the MkDocs-material theme will "stick" to something depending on the time of day the user first came to it, which seems arbitrary.

For posterity—who may suffer similar confusion and end up in this thread trying match dark/light images to dark/light themes—I added two (somewhat fragile) work-arounds to [my example](https://posita.github.io/mkdocs-dark-bug/):

1. Use `extra_css` to tie dark/light image selection solely to the user-selected-or-toggled theme ([here](https://posita.github.io/mkdocs-dark-bug/latest/#work-around-using-extra_css-to-tie-image-selection-to-user-selected-theme)).
1. Use `extra_javascript` to tie image and theme selection to independently react to changes to `prefers-color-scheme` ([here](https://posita.github.io/mkdocs-dark-bug/latest/#work-around-using-extra_javascript-to-tie-theme-selection-to-prefers-color-scheme)). In this case, you probably want to disable users from selecting/toggling themes.

**Comment by squidfunk** [comment]
I'm happy to announce that automatic dark mode was finally implemented and just released as part of [8.3.4+insiders-4.18.0.](https://squidfunk.github.io/mkdocs-material/insiders/changelog/#4.18.0) I invite you to test-drive it on [our documentation](https://squidfunk.github.io/mkdocs-material/). Here's a short video showing how it works:

https://user-images.githubusercontent.com/932156/173182190-f5f9ab50-1a1e-414d-8184-530843f3362f.mp4
